llvm_mode/compare-transform-pass.so.cc

```diff
@@ -304,17 +304,27 @@ bool CompareTransform::transformCmps(Module &M, const bool processStrcmp,
           if (!(HasStr1 || HasStr2)) continue;
 
           if (isMemcmp || isStrncmp || isStrncasecmp) {
-
             /* check if third operand is a constant integer
              * strlen("constStr") and sizeof() are treated as constant */
             Value *      op2 = callInst->getArgOperand(2);
             ConstantInt *ilen = dyn_cast<ConstantInt>(op2);
-            if (!ilen) continue;
-            /* final precaution: if size of compare is larger than constant
-             * string skip it*/
-            uint64_t literalLength = HasStr1 ? Str1.size() : Str2.size();
-            if (literalLength + 1 < ilen->getZExtValue()) continue;
-
+            if (ilen) {
+              uint64_t len = ilen->getZExtValue();
+              // if len is zero this is a pointless call but allow real
+              // implementation to worry about that
+              if (!len) continue;
+
+              if (isMemcmp) {
```

Comment:
why do you limit this check to memcmp? just a curious question

Reply:
Because in string comparison it's perfectly valid - if you supply an `n` greater than the const string's length the comparison will still terminate at the end of the const string because either

 - the variable string won't have a null char in that position or
 - the variable string _does_ have a null char in that position, but of course that denotes the end of the variable string and so the comparison should terminate.

This might be a perfectly reasonable thing to do if e.g. the code just uses the var string's allocation length for `n` to guard against missing terminators.

In the `memcmp` case, the presence of a null char won't terminate the comparison and the expected behaviour would be to go on comparing past the end of the const string. I'd hope the compiler or `FORTIFY` flags would catch this kind of thing but I didn't look too far into it.

Reply:
true